k2/csrc/fsa_algo.cc

```diff
@@ -147,4 +147,109 @@ void Connect(const Fsa &a, Fsa *b, std::vector<int32_t> *arc_map /*=nullptr*/) {
   }
 }
 
+bool TopoSort(const Fsa &a, Fsa *b,
+              std::vector<int32_t> *state_map /*= nullptr*/) {
+  CHECK_NOTNULL(b);
+  b->arc_indexes.clear();
+  b->arcs.clear();
+
+  if (state_map) state_map->clear();
+
+  if (IsEmpty(a)) return false;
+  static constexpr int32_t kNotVisited = 0;  // a node that has not been visited
+  static constexpr int32_t kVisiting = 1;    // a node that is under visiting
+  static constexpr int32_t kVisited = 2;     // a node that has been visited
+
+  auto num_states = a.NumStates();
+  auto final_state = num_states - 1;
+  std::vector<int8_t> state_status(num_states, kNotVisited);
+
+  // map order to state.
+  // state 0 has the largest order, i.e., num_states - 1
+  // final_state has the least order, i.e., 0
+  std::vector<int32_t> order;
+
```

Comment:
also remove this blank line, as you `reserve` it just after defining it.

Reply:
Done.